versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/MergeBase.java

```diff
@@ -0,0 +1,324 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.agrona.collections.Object2ObjectHashMap;
+import org.immutables.value.Value;
+import org.projectnessie.versioned.storage.common.objtypes.CommitObj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+
+/**
+ * Identifies the nearest commit that shall be used as the base commit for a merge of the given
+ * {@link #fromCommitId() from-commit} onto the given {@link #targetCommitId() target commit}.
+ *
+ * <p>This class also supports finding the base commit for N-way merges, although N-way merges are
+ * not implemented for Nessie (yet).
+ */
+@Value.Immutable
+public abstract class MergeBase {
+  public abstract Function<ObjId, CommitObj> loadCommit();
+
+  public abstract ObjId targetCommitId();
+
+  public abstract ObjId fromCommitId();
+
+  /**
+   * Whether merge-parents shall be respected, defaults to {@code true}. Setting this to {@code
+   * false} changes the behavior to return the <em>common ancestor</em> instead of the nearest
+   * merge-base.
+   */
+  @Value.Default
+  public boolean respectMergeParents() {
+    return true;
+  }
+
+  public static ImmutableMergeBase.Builder builder() {
+    return ImmutableMergeBase.builder();
+  }
+
+  @Value.NonAttribute
+  public ObjId identifyMergeBase() {
+    List<ShallowCommit> mergeBases = identifyAllMergeBases();
+    if (mergeBases == null || mergeBases.isEmpty()) {
+      throw noCommonAncestor();
+    }
+    return mergeBases.get(0).id();
+  }
+
+  private List<ShallowCommit> identifyAllMergeBases() {
+    ShallowCommit targetCommit = shallowCommit(targetCommitId());
+    if (targetCommit == null) {
+      // This is rather a UX hack, to raise a "not found" exception if any of the from-commit-IDs
+      // does not exist
+      shallowCommit(fromCommitId());
+      return null;
+    }
+
+    ShallowCommit fromCommit = shallowCommit(fromCommitId());
+    if (fromCommit == null) {
+      return singletonList(targetCommit);
+    }
+
+    return findMergeBases(fromCommit, targetCommit);
+  }
+
+  private List<ShallowCommit> findMergeBases(ShallowCommit commitA, ShallowCommit commitB) {
+    if (commitB.id().equals(commitA.id())) {
+      return newArrayList(commitA);
+    }
+
+    List<ShallowCommit> reachableCommits = flagReachableCommits(commitA, commitB);
+
+    clearFlags(newArrayList(commitA, commitB), ALL_FLAGS);
+
+    return removeRedundant(reachableCommits);
+  }
+
+  private List<ShallowCommit> flagReachableCommits(ShallowCommit commitA, ShallowCommit commitB) {
+    PriorityQueue<ShallowCommit> queue = new PriorityQueue<>(comparing(ShallowCommit::seq));
+    List<ShallowCommit> result = new ArrayList<>();
+
+    commitA.setCommitA();
+    queue.add(commitA);
+
+    commitB.setCommitB();
+    queue.add(commitB);
+
+    while (queue.stream().anyMatch(ShallowCommit::isNotCandidate)) {
+      ShallowCommit commit = requireNonNull(queue.poll());
```

Comment:
I'm not sure the iteraction order is ideal here. This is what I observed for `TestMergeBase.multiMerge2()`:

```
commit ID: 00000006, seq: 6
commit ID: 00000003, seq: 3
commit ID: 00000000, seq: 0
commit ID: 00000005, seq: 5
commit ID: 00000004, seq: 4
commit ID: 00000002, seq: 2
commit ID: 00000001, seq: 1
commit ID: 00000008, seq: 8
commit ID: 00000007, seq: 7
commit ID: 00000001, seq: 1
commit ID: 00000000, seq: 0
commit ID: 00000003, seq: 3
```

I think it'd be preferable to process from more recent commits to older commits (i.e. reversing the priority queue's comparator). Then, perhaps we won't need the extra `removeRedundant` step?.. WDYT?

Reply:
Also, current iteration strategy will always go to the root, it seems, but we should be able to stop as soon as we reach the deepest merge base (which is not root in most cases).

Reply:
It can "shoot over the common ancestor" a little bit, but won't traverse the whole history.